tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc

```diff
@@ -1644,6 +1648,90 @@ TEST_F(OpConverterTest, ConvertReshape) {
   }
 }
 
+// Helper function for testing MatMul and BatchMatMul
+// get_matmul corresponds to the function used to generate the node. It should
+// accept (DataType, transpose_a, transpose_b) as parameters.
+void OpConverterTest::TestMatMulHelper(
+    const std::function<NodeDef(DataType, bool, bool)>& get_matmul,
+    const std::string& op_name) {
+  // HACK: This needs to be done in a better way.
+  const bool is_batch_matmul = op_name == "BatchMatMul";
+  {
+    // Unsupported data type.
+    Reset();
+    NodeDef node_def = get_matmul(DT_INT32, false, false);
+    AddTestTensor("input", {2}, /*batch_size=*/1, nvinfer1::DataType::kINT32);
+    AddTestWeights<int32>("weights", {2, 1}, {3, 5});
+    RunValidationAndConversion(
+        node_def, error::UNIMPLEMENTED,
+        ("Data type int32 is not supported for " + op_name +
+         ", "
+         "must be one of [float, half], at my_matmul")
+            .c_str());
+  }
+  // OK.
+  for (bool transpose_a : {false, true}) {
+    for (bool transpose_b : {false, true}) {
+      Reset();
+      NodeDef node_def = get_matmul(DT_FLOAT, transpose_a, transpose_b);
+      AddTestTensor("input", {2}, /*batch_size=*/1);
+      AddTestWeights<float>("weights", {2, 2}, {0, 1, 2, 3});
+      if (is_batch_matmul) {
+        RunValidationAndConversion(
+            node_def, error::INVALID_ARGUMENT,
+            "Input weight attempts to broadcast across batch dimension");
+        continue;
+      } else if (transpose_a) {
+        RunValidationAndConversion(
+            node_def, error::INVALID_ARGUMENT,
+            "Cannot transpose first input if it is a tensor with fewer than 2 "
+            "non-batch dimensions");
+        continue;
+      }
+      RunValidationAndConversion(node_def);
+      TRT_TensorOrWeights output;
+      TF_EXPECT_OK(GetTensorOrWeights("my_matmul", &output));
+      ASSERT_TRUE(output.is_tensor());
+      ExpectTrtDimsEqualsArray({2}, output.tensor()->getDimensions());
+
+      const DataVec input_data{{"input", test::AsTensor<float>({0, 1})}};
+      DataVec output_data{{"my_matmul", ConstructTensor<float>(2)}};
+      BuildAndRun(input_data, &output_data);
+      if (transpose_b) {
+        EXPECT_THAT(GetSpanForData<float>(output_data[0]), ElementsAre(1, 3));
+      } else {
+        EXPECT_THAT(GetSpanForData<float>(output_data[0]), ElementsAre(2, 3));
+      }
+    }
+  }
+  // OK, 3D inputs
+  for (bool transpose_b : {false, true}) {
+    Reset();
+    NodeDef node_def = get_matmul(DT_FLOAT, /*transpose_a=*/false, transpose_b);
+    AddTestTensor("input", {1, 1, 2}, /*batch_size=*/1);
+    AddTestWeights<float>("weights", {2, 2}, {0, 1, 2, 3});
```

Comment:
I don't think TF MatMul op supports shape(1,1,1,2)*shape(2,2), batchmatmulv2 will support that though. I think MatMul requires that both inputs are real 2D matrix.

Also, it's unfortunate that our test suite doesn't run TF kernel. I'll fix that soon (but probably not before this PR is merged :( )..

Reply:
So here you probably want to do: shape(A)=[1,2,3], shape(B)=[1,3,2], and output_shape=[1,2,2] (only test batchmatmul)

Reply:
From the docs it seems like >= 2D is supported https://www.tensorflow.org/api_docs/python/tf/linalg/matmul

I'll add the additional test case for BatchMatMul

Reply:
tf.matmul supports >2D because it'll use op batchmatmul to handle them. Also note there is no tf.batch_matmul, and tf.matmul decides to use matmul or batchmatmul